### ISSUE TITLE
SpMV out-of-bounds accesses

### ISSUE THREAD

**Issue by marsaev** [NVIDIA/cub#139]
Reproduces with any input.

```c++
$ cat spmvtest.cu
#include <stdio.h>
#include <stdlib.h>
#include "cub/cub.cuh"

#if 0
#define CUCHK(call) (call);
#else
#define CUCHK(call) {                                    \
    cudaError_t err = call;                                                    \
    if( cudaSuccess != err) {                                                \
      fprintf(stderr, "Cuda error in file '%s' in line %i : %s.\n",        \
              __FILE__, __LINE__, cudaGetErrorString( err) );              \
      fflush(stderr); \
      exit(EXIT_FAILURE);                                                  \
    } }
#endif

int main(int argc, char *argv[])
{
  int ret_code;
  int M = 5, N = 5, nz = 5;
  double alpha = 1, beta = 0;

  int colIdx[5] = {0, 1, 2, 3, 4};
  int rowOff[6] = {0, 1, 2, 3, 4, 5};
  double val[5] = { 1., 1., 1., 1., 1. };
  double xval[5] = { 1., 1., 1., 1., 1. };
  double out[5] = { 0., 0., 0., 0., 0. };
  double dout[5];


  int *d_csrRow, *d_col;
  double *d_val, *d_x, *d_y;
  void *tmp;
  CUCHK(cudaMalloc(&d_csrRow, (N+1)*sizeof(int)));
  CUCHK(cudaMalloc(&d_col, nz*sizeof(int)));
  CUCHK(cudaMalloc(&d_val, nz*sizeof(double)));
  CUCHK(cudaMalloc(&d_x, N*sizeof(double)));
  CUCHK(cudaMalloc(&d_y, N*sizeof(double)));

  CUCHK(cudaMemcpy(d_csrRow, rowOff, (N+1)*sizeof(int), cudaMemcpyHostToDevice));
  CUCHK(cudaMemcpy(d_col, colIdx, nz*sizeof(int), cudaMemcpyHostToDevice));
  CUCHK(cudaMemcpy(d_val, val, nz*sizeof(double), cudaMemcpyHostToDevice));
  CUCHK(cudaMemcpy(d_x, xval, N*sizeof(double), cudaMemcpyHostToDevice));
  CUCHK(cudaMemcpy(d_y, out, N*sizeof(double), cudaMemcpyHostToDevice));
  CUCHK(cudaGetLastError());

  size_t tsize;

  cub::DeviceSpmv::CsrMV(NULL, tsize, d_val,
    d_csrRow, d_col, d_x, d_y,
    M, N, nz);

  CUCHK(cudaMalloc(&tmp, tsize));

  cub::DeviceSpmv::CsrMV(tmp, tsize, d_val,
    d_csrRow, d_col, d_x, d_y,
    M, N, nz);

  CUCHK(cudaDeviceSynchronize());
  CUCHK(cudaMemcpy(dout, d_y, N*sizeof(double), cudaMemcpyDeviceToHost));

  // verify GPU results
  for (int i = 0; i < N; i++) {
    out[i] *= beta;
    for (int j = rowOff[i]; j < rowOff[i+1]; j++) {
      out[i] += alpha*val[j]*xval[colIdx[j]];
    }
  }
  double err = 0, norm = 0;
  for (int i = 0; i < N; i++) {
    err += fabs(out[i] - dout[i]);
    norm += out[i];
  }

  printf("err = %g, norm = %g, err/norm = %g\n", err, norm, err/norm);

  cudaFree(d_csrRow);
  cudaFree(d_col);
  cudaFree(d_val);
  cudaFree(d_x);
  cudaFree(d_y);
}
```
Error:
```c++
$ /usr/local/cuda-9.2/bin/nvcc spmvtest.cu -Icub
$ /usr/local/cuda-9.2/bin/cuda-memcheck ./a.out
========= CUDA-MEMCHECK
========= Invalid __global__ read of size 4
=========     at 0x00000950 in void cub::DeviceSpmvKernel<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, cub::ReduceByKeyScanTileState<double, int, bool=1>, double, int, int2, bool=0, bool=0>(cub::SpmvParams<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, double>, int*, cub::KeyValuePair<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, cub::SpmvParams>*, int, int, int)
=========     by thread (5,0,0) in block (0,0,0)
=========     Address 0x7f2468800018 is out of bounds
=========     Device Frame:void cub::DeviceSpmvKernel<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, cub::ReduceByKeyScanTileState<double, int, bool=1>, double, int, int2, bool=0, bool=0>(cub::SpmvParams<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, double>, int*, cub::KeyValuePair<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, cub::SpmvParams>*, int, int, int) (void cub::DeviceSpmvKernel<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, cub::ReduceByKeyScanTileState<double, int, bool=1>, double, int, int2, bool=0, bool=0>(cub::SpmvParams<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, double>, int*, cub::KeyValuePair<cub::DispatchSpmv<double, int>::PtxSpmvPolicyT, cub::SpmvParams>*, int, int, int) : 0x950)
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so.1 (cuLaunchKernel + 0x2c5) [0x240035]
=========     Host Frame:./a.out [0x1e502]
=========     Host Frame:./a.out [0x1e6f7]
=========     Host Frame:./a.out [0x4d485]
=========     Host Frame:./a.out [0x796c]
=========     Host Frame:./a.out [0x7463]
=========     Host Frame:./a.out [0x74d8]
=========     Host Frame:./a.out [0x9174]
=========     Host Frame:./a.out [0x8e15]
=========     Host Frame:./a.out [0x6b23]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xe7) [0x21b97]
=========     Host Frame:./a.out [0x5f5a]
=========
Cuda error in file 'spmvtest.cu' in line 60 : unspecified launch failure.
========= Program hit cudaErrorLaunchFailure (error 4) due to "unspecified launch failure" on CUDA API call to cudaDeviceSynchronize.
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so.1 [0x32a783]
=========     Host Frame:./a.out [0x3ae46]
=========     Host Frame:./a.out [0x6b2c]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xe7) [0x21b97]
=========     Host Frame:./a.out [0x5f5a]
=========
========= ERROR SUMMARY: 2 errors
```

Happens in this line when reading row offset for the last item in last tile: https://github.com/NVlabs/cub/blob/c3cceac115c072fb63df1836ff46d8c60d9eb304/cub/agent/agent_spmv_orig.cuh#L487

And just for the information - I tried to add workaround for the read and then there is also write OOB in the fixup kernel.

**Comment by afender** [comment]
This repro in several analytics functions in RAPIDS/cugraph. 
Saw it on DGX1 with cuda 10.1.

**Comment by allisonvacanti** [comment]
#160 points to the location of the error.

**Comment by allisonvacanti** [comment]
This will be fixed by #352.
